astropy/io/fits/fitsrec.py

```diff
@@ -1283,19 +1284,50 @@ def _get_recarray_field(array, key):
     return field
 
 
-def _rstrip_inplace(array, chars=None):
+def _rstrip_inplace(array):
     """
-    Performs an in-place rstrip operation on string arrays.
-    This is necessary since the built-in `np.char.rstrip` in Numpy does not
-    perform an in-place calculation.  This can be removed if ever
-    https://github.com/numpy/numpy/issues/6303 is implemented (however, for
-    the purposes of this module the only in-place vectorized string functions
-    we need are rstrip and encode).
+    Performs an in-place rstrip operation on string arrays. This is necessary
+    since the built-in `np.char.rstrip` in Numpy does not perform an in-place
+    calculation.
     """
 
-    for item in np.nditer(array, flags=['zerosize_ok'],
-                                 op_flags=['readwrite']):
-        item[...] = item.item().rstrip(chars)
+    # The following implementation convert the string to unsigned integers of
+    # the right length. Trailing spaces (which are represented as 32) are then
+    # converted to null characters (represented as zeros). To avoid creating
+    # large temporary mask arrays, we loop over chunks (attempting to do that
+    # on a 1-D version of the array; large memory may still be needed in the
+    # unlikely case that a string array has small first dimension and cannot
+    # be represented as a contiguous 1-D array in memory).
+
+    dt = array.dtype
+
+    if dt.kind not in 'SU':
+        raise TypeError("This function can only be used on string arrays")
+    # View the array as appropriate integers. The last dimension will
+    # equal the number of characters in each string.
+    bpc = 1 if dt.kind == 'S' else 4
+    dt_int = "{0}{1}u{2}".format(dt.itemsize // bpc, dt.byteorder, bpc)
+    b = np.array(array, copy=False).view(dt_int)
```

Comment:
Actually, why are we doing the `np.array` here? If it is not a proper array, this will copy and nothing will happen. Maybe replace with `b = array.view(dt_int, np.ndarray)` (it is useful, I think, so remove any subclass here)

Reply:
Because array is a ``chararray`` and things don't work otherwise...

Reply:
Specifically:

```
In [12]: array = np.array([b'a ', b'b ', b'c ']).view(np.chararray)

In [13]: array.view('2|u1')
Out[13]: ---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
~/miniconda3/envs/dev/lib/python3.6/site-packages/IPython/core/formatters.py in __call__(self, obj)
    691                 type_pprinters=self.type_printers,
    692                 deferred_pprinters=self.deferred_printers)
--> 693             printer.pretty(obj)
    694             printer.flush()
    695             return stream.getvalue()

~/miniconda3/envs/dev/lib/python3.6/site-packages/IPython/lib/pretty.py in pretty(self, obj)
    378                             if callable(meth):
    379                                 return meth(obj, self, cycle)
--> 380             return _default_pprint(obj, self, cycle)
    381         finally:
    382             self.end_group()

~/miniconda3/envs/dev/lib/python3.6/site-packages/IPython/lib/pretty.py in _default_pprint(obj, p, cycle)
    493     if _safe_getattr(klass, '__repr__', None) is not object.__repr__:
    494         # A user-provided repr. Find newlines and replace them with p.break_()
--> 495         _repr_pprint(obj, p, cycle)
    496         return
    497     p.begin_group(1, '<')

~/miniconda3/envs/dev/lib/python3.6/site-packages/IPython/lib/pretty.py in _repr_pprint(obj, p, cycle)
    691     """A pprint that just redirects to the normal repr function."""
    692     # Find newlines and replace them with p.break_()
--> 693     output = repr(obj)
    694     for idx,output_line in enumerate(output.splitlines()):
    695         if idx:

~/miniconda3/envs/dev/lib/python3.6/site-packages/numpy/core/numeric.py in array_repr(arr, max_line_width, precision, suppress_small)
   1896     if arr.size > 0 or arr.shape == (0,):
   1897         lst = array2string(arr, max_line_width, precision, suppress_small,
-> 1898                            ', ', "array(")
   1899     else:  # show zero-length shape unless it is (0,)
   1900         lst = "[], shape=%s" % (repr(arr.shape),)

~/miniconda3/envs/dev/lib/python3.6/site-packages/numpy/core/arrayprint.py in array2string(a, max_line_width, precision, suppress_small, separator, prefix, style, formatter)
    461     else:
    462         lst = _array2string(a, max_line_width, precision, suppress_small,
--> 463                             separator, prefix, formatter=formatter)
    464     return lst
    465 

~/miniconda3/envs/dev/lib/python3.6/site-packages/numpy/core/arrayprint.py in _array2string(a, max_line_width, precision, suppress_small, separator, prefix, formatter)
    334     lst = _formatArray(a, format_function, len(a.shape), max_line_width,
    335                        next_line_prefix, separator,
--> 336                        _summaryEdgeItems, summary_insert)[:-1]
    337     return lst
    338 

~/miniconda3/envs/dev/lib/python3.6/site-packages/numpy/core/arrayprint.py in _formatArray(a, format_function, rank, max_line_len, next_line_prefix, separator, edge_items, summary_insert)
    529             if leading_items or i != trailing_items:
    530                 s += next_line_prefix
--> 531             s += _formatArray(a[-i], format_function, rank-1, max_line_len,
    532                               " " + next_line_prefix, separator, edge_items,
    533                               summary_insert)

~/miniconda3/envs/dev/lib/python3.6/site-packages/numpy/core/defchararray.py in __getitem__(self, obj)
   1850 
   1851     def __getitem__(self, obj):
-> 1852         val = ndarray.__getitem__(self, obj)
   1853 
   1854         if isinstance(val, character):

~/miniconda3/envs/dev/lib/python3.6/site-packages/numpy/core/defchararray.py in __array_finalize__(self, obj)
   1847         # The b is a special case because it is used for reconstructing.
   1848         if not _globalvar and self.dtype.char not in 'SUbc':
-> 1849             raise ValueError("Can only create a chararray from string data.")
   1850 
   1851     def __getitem__(self, obj):

ValueError: Can only create a chararray from string data.
```

Reply:
OK, but then we still should just do a `.view(dt_int, np.ndarray)` - that way we get a failure rather than a copy for the wrong input.

Reply:
ps. The view does work if you also pass in the class.

Reply:
Ah I see, I didn't know about that trick!

Reply:
It doesn't seem to work for chararray unfortunately. However I changed it so that I only do the explicit call to np.array if the array is a chararray, which should avoid accidental copies I think.

Reply:
That's weird; I had checked:
```
In [8]: np.char.array(['abc', 'bcd']).view('<3u4', np.ndarray)
Out[8]: 
array([[ 97,  98,  99],
       [ 98,  99, 100]], dtype=uint32)
```
Might it be version-dependent?

Reply:
Ah I think I got it to work now, not sure what was happening before :-/